File: lab1/keytoktab.c
/**********************************************************************/
/*Per Emilsson och Kenny Pettersson                                   */
/**********************************************************************/

/**********************************************************************/
/* lab 1 DVG C01 - Driver OBJECT                                      */
/**********************************************************************/

/**********************************************************************/
/* Include files                                                      */
/**********************************************************************/
#include <stdio.h>
#include <ctype.h>
#include <string.h>

/**********************************************************************/
/* Other OBJECT's METHODS (IMPORTED)                                  */
/**********************************************************************/
#include "keytoktab.h"

/**********************************************************************/
/* OBJECT ATTRIBUTES FOR THIS OBJECT (C MODULE)                       */
/**********************************************************************/
/**********************************************************************/
/* type definitions                                                   */
/**********************************************************************/
typedef struct tab {
	char 	* text;
	int 	token;
}tab;

/**********************************************************************/
/* data objects (tables)                                              */
/**********************************************************************/
static tab tokentab[ ] = {
    {"id", 	            id},
    {"number",      number},
    {":=", 	        assign},
    {"undef", 	     undef},
    {"predef",      predef},
    {"tempty",      tempty},
    {"error",        error},
    {"type",           typ},
    {"$",              '$'},
    {"(",              '('},
    {")",              ')'},
    {"*",              '*'},
    {"+",              '+'},
    {",",              ','},
    {"-",              '-'},
    {".",              '.'},
    {"/",              '/'},
    {":",              ':'},
    {";",              ';'},
    {"=",              '='},
    {"TERROR", 	    nfound}
};


static tab keywordtab[ ] = {
	{"program", 	program},
	{"input", 	      input},
	{"output", 	     output},
	{"var", 	        var},
	{"begin", 	      begin},
	{"end", 	        end},
	{"boolean", 	boolean},
	{"integer", 	integer},
	{"real", 	       real},
	{"KERROR", 	     nfound}
};

/**********************************************************************/
/*  PUBLIC METHODS for this OBJECT  (EXPORTED)                        */
/**********************************************************************/
/**********************************************************************/
/* Display the tables                                                 */
/**********************************************************************/
void p_toktab()
{
    printf("\nTHE PROGRAM KEYWORDS");
    for(int i = 0; i < (sizeof(keywordtab) / sizeof(keywordtab[0])); i++){
        printf("\n%-12s %4d", keywordtab[i].text, keywordtab[i].token);
    }
    
    printf("\n\nTHE PROGRAM TOKENS");
    for(int i = 0; i < (sizeof(tokentab) / sizeof(tokentab[0])); i++){
        printf("\n%-12s %4d", tokentab[i].text, tokentab[i].token);
    }
}

/**********************************************************************/
/* lex2tok - convert a lexeme to a token                              */
/**********************************************************************/
toktyp lex2tok(char * fplex)
{
    for(int i = 0; i < (sizeof(tokentab) / sizeof(tokentab[0])); i++){
        if(strcmp(tokentab[i].text, fplex) == 0){
            return tokentab[i].token;
        }
    }
    
    for(int i = 0; i < (sizeof(keywordtab) / sizeof(keywordtab[0])); i++){
        if(strcmp(keywordtab[i].text, fplex) == 0){
            return keywordtab[i].token;
        }
    }
	//handle digits
    if(isdigit((int)*fplex)) {
		return tokentab[1].token;
	}
	return tokentab[0].token;
}

/**********************************************************************/
/* key2tok - convert a keyword to a token                             */
/**********************************************************************/
toktyp key2tok(char * fplex)
{ 
    for(int i = 0; i < (sizeof(keywordtab) / sizeof(keywordtab[0])); i++){
        if(strcmp(keywordtab[i].text, fplex) == 0){
            return keywordtab[i].token;
        }
    }	

    return tokentab[0].token;
}

/**********************************************************************/
/* tok2lex - convert a token to a lexeme                              */
/**********************************************************************/
char * tok2lex(toktyp ftok)
{
    for(int i = 0; i < (sizeof(tokentab) / sizeof(tokentab[0])); i++){
        if(tokentab[i].token == ftok){
            return tokentab[i].text;
        }
    }

    for(int i = 0; i < (sizeof(keywordtab) / sizeof(keywordtab[0])); i++){
        if(keywordtab[i].token == ftok){
            return keywordtab[i].text;
        }
    }
    
    return tokentab[0].text;
}

/**********************************************************************/
/* End of code                                                        */
/**********************************************************************/

File: lab4/keytoktab.py
from enum import IntEnum
from dataclasses import dataclass
from typing import List, Union, Dict

NFOUND = -1

class toktyp(IntEnum):
    START       = 257
    ID          = 258
    NUMBER      = 259
    ASSIGN      = 260
    PREDEF      = 261
    TEMPTY      = 262
    UNDEF       = 263
    ERROR       = 264
    TYPE        = 265
    END         = 266
    KSTART      = 267
    PROGRAM     = 268
    INPUT       = 269
    OUTPUT      = 270
    VAR         = 271
    BEGIN       = 272
    END_KW      = 273
    BOOLEAN     = 274
    INTEGER     = 275
    REAL        = 276
    KEND        = 277

@dataclass() #generates __init__ etc
class tab:
    text: str
    token: Union[toktyp, str] #can hold either a toktyp or a string

tokentab: List[tab] = [
    tab("id",      toktyp.ID),
    tab("number",  toktyp.NUMBER),
    tab(":=",      toktyp.ASSIGN),
    tab("undef",   toktyp.UNDEF),
    tab("predef",  toktyp.PREDEF),
    tab("tempty",  toktyp.TEMPTY),
    tab("error",   toktyp.ERROR),
    tab("type",    toktyp.TYPE),
    tab("$",       '$'),
    tab("(",       '('),
    tab(")",       ')'),
    tab("*",       '*'),
    tab("+",       '+'),
    tab(",",       ','),
    tab("-",       '-'),
    tab(".",       '.'),
    tab("/",       '/'),
    tab(":",       ':'),
    tab(";",       ';'),
    tab("=",       '='),
    tab("TERROR",  NFOUND),
]

keywordtab: List[tab] = [
    tab("program",  toktyp.PROGRAM),
    tab("input",    toktyp.INPUT),
    tab("output",   toktyp.OUTPUT),
    tab("var",      toktyp.VAR),
    tab("begin",    toktyp.BEGIN),
    tab("end",      toktyp.END_KW),
    tab("boolean",  toktyp.BOOLEAN),
    tab("integer",  toktyp.INTEGER),
    tab("real",     toktyp.REAL),
    tab("KERROR",   NFOUND),
]

# build a single dict for all of them
token_map: Dict[str, Union[toktyp, str]] = { entry.text: entry.token for entry in tokentab + keywordtab }

#display the tables
def p_toktab():
    print("\nTHE PROGRAM KEYWORDS")
    for entry in keywordtab:
        print(f"{entry.text:<12} {entry.token:4}")

    print("\n\nTHE PROGRAM TOKENS")
    for entry in tokentab:
        print(f"{entry.text:<12} {entry.token:4}")

#lex2tok - convert a lexeme to a token 
def lex2tok(lexeme: str) -> Union[toktyp, str]:
    
    #use pythons isidentifier() to check if it is an ID and not in the token_map dict
    if lexeme.isidentifier() and lexeme not in token_map:
        return toktyp.ID
        
    if lexeme.isdigit():
        return toktyp.NUMBER

    # otherwise use the get() function for Dicts that checks if the lexeme is in the dict otherwise returns error
    return token_map.get(lexeme, toktyp.ERROR)
    
#key2tok - convert a keyword to a token
def key2tok(lexeme: str) -> Union[toktyp, str]:

    #we use pythons next() function to scan through all keywords and return the matching one or else ERROR
    for entry in keywordtab:
        if entry.text == lexeme:
            return entry.token
    
    # default return error
    return toktyp.ERROR

#tok2lex - convert a token to a lexeme  
def tok2lex(ftok: Union[toktyp, str]) -> str:
    for text, tok in token_map.items():
        if tok == ftok:
            return text
    return "error"